docs/runtime/ir-caching.md

```diff
@@ -0,0 +1,209 @@
+In order to maximise the benefits of this process, we want to serialise the IR
+as _late_ in the compiler pipeline as possible. This means serialising it just
+before the code generation step that generates Truffle nodes (before the
+`RuntimeStubsGenerator` and `IrToTruffle` run).
+
+This serialisation should take place in an _offloaded thread_ so that
+serialisation doesn't block codegen from occurring.
+
+### Breaking Links
+
+Doing this naïvely, however, means that we can inadvertently end up serialising
+the entire module graph. This is due to the `BindingsMap`, which contains a
+reference to the associated `runtime.Module`, from which there is a reference to
+the `ModuleScope`. The `ModuleScope` may then reference other `runtime.Module`s
+which all contain `IR.Module`s. Therefore, done in a silly fashion, we end up
+serialising the entire reachable module graph. This is not what we want.
+
+While the ideal way of solving this problem would be to customise the
+serialisation and deserialisation process for the `BindingsMap`, the JVM's
+`Serializable` does not provide the ability to customise it enough to solve this
+problem. Instead, we solve it using a preprocessing step:
+
+- We can modify `BindingsMap` and its child types to be able to contain an
+  unlinked module pointer
+  `case class ModulePointer(namespace: String, qualifiedName: List[String])` in
+  place of a `Module`.
+- As the `MetadataStorage` type that holds the `BindingsMap` is mutable it can
+  be updated in place without having to reassemble the entire IR graph.
+- Hence, we can traverse all the nodes in the `ir.preorder` that have metadata
+  consisting of either the `BindingsMap` or `ResolvedName` types (provided by
+  the following passes: `BindingAnalysis`, `MethodDefinitions`,
+  `UppercaseNames`, `VectorLiterals`, `Patterns`), and perform a replacement.
+
+Having done this, we have broken any links that the IR may hold between modules,
+and can serialise each module individually.
+
+## Storing the IR
+
+The serialized IR needs to be stored in a location that is tied to the library
+that it serializes. Despite this, we _also_ want to be able to ship cached IR
+with libraries. This leads to a two pronged solution where we check two
+locations for the cache.
+
+1. **With the Library:** As libraries can have a hidden `.enso` directory, we
+   can use a path within that for caching. This should be
+   `$package/.enso/cache/ir/enso-$version/`, and can be got at by extending the
+   `pkg` library to be aware of the cache directories.
+2. **Globally:** As some library locations may not be writable, we need to have
+   a global out-of-line cache that is used if the first one is not writable.
+   This is located under `$ENSO_DATA` (whose location can be obtained from the
+   `RuntimeDistributionManager`), and is located under the path
+   `$ENSO_DATA/cache/ir/$hash/enso-$version/`, where `$hash` is the `SHA3-224`
+   hash of the tuple `(namespace, library_name, version)`, where
```

Comment:
While this may not be crucial for this documentation, as this is rather a high-level overview, it is unclear to me how this hash is computed - are the string representations of each of this values concatenated, or what exactly? I'm not sure if it _has_ to be clarified here, just wondering, as the current description is not completely clear about this detail.